src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSObject.References.cs

```diff
@@ -10,53 +10,37 @@ namespace System.Runtime.InteropServices.JavaScript
     public partial class JSObject
     {
         internal nint JSHandle;
-
-#if FEATURE_WASM_THREADS
-        private readonly object _thisLock = new object();
-        private SynchronizationContext? m_SynchronizationContext;
-#endif
+        internal JSProxyContext ProxyContext;
 
         public SynchronizationContext SynchronizationContext
         {
             get
             {
 #if FEATURE_WASM_THREADS
-                return m_SynchronizationContext!;
+                return ProxyContext.SynchronizationContext;
 #else
                 throw new PlatformNotSupportedException();
 #endif
             }
         }
 
-#if FEATURE_WASM_THREADS
-        // the JavaScript object could only exist on the single web worker and can't migrate to other workers
-        internal nint OwnerTID;
-#endif
 #if !DISABLE_LEGACY_JS_INTEROP
         internal GCHandle? InFlight;
         internal int InFlightCounter;
 #endif
-        private bool _isDisposed;
+        internal bool _isDisposed;
 
-        internal JSObject(IntPtr jsHandle)
+        internal JSObject(IntPtr jsHandle, JSProxyContext ctx)
         {
+            ProxyContext = ctx;
             JSHandle = jsHandle;
-#if FEATURE_WASM_THREADS
-            var ctx = JSSynchronizationContext.CurrentJSSynchronizationContext;
-            if (ctx == null)
-            {
-                Environment.FailFast("Missing CurrentJSSynchronizationContext");
-            }
-            m_SynchronizationContext = ctx;
-            OwnerTID = ctx!.TargetTID;
-#endif
         }
 
 #if !DISABLE_LEGACY_JS_INTEROP
         internal void AddInFlight()
         {
             ObjectDisposedException.ThrowIf(IsDisposed, this);
-            lock (this)
+            lock (ProxyContext)
```

Comment:
if AddInFlight is called from a different thread, this now serializes the thread calling AddInFlight with the thread that owns the object. I don't know if that's a problem or not. It's probably fine

Reply:
I will delete whole legacy interop and this will go too. 
Also legacy interop is ST/UI only.